certora/specs/MintBurnEthereum.spec

```diff
@@ -0,0 +1,135 @@
+*/
+rule noChangeTotalSupply(env e, method f) filtered {
+    f-> f.selector != sig:upgradeToAndCall(address, bytes).selector
+} {
+    requireInvariant totalSupplyIsSumOfBalances();
+    requireInvariant balancesLTEqTotalSupply();
+
+    calldataarg args;
+
+    uint256 totalSupplyBefore = totalSupply();
+    f(e, args);
+    uint256 totalSupplyAfter = totalSupply();
+
+    assert totalSupplyAfter > totalSupplyBefore => f.selector == sig:mint(address,uint256).selector || f.selector == sig:initialize(address, address).selector;
+    assert totalSupplyAfter < totalSupplyBefore => f.selector == sig:burn(uint256).selector;
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rules: mint behavior and side effects                                                                               │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule mint(env e) {
+    requireInvariant totalSupplyIsSumOfBalances();
+    requireInvariant balancesLTEqTotalSupply();
+    assert isTotalSupplyGTEqSumOfVotingPower();
+    requireInvariant zeroAddressNoVotingPower();
+    require nonpayable(e);
+
+    address to;
+    address other;
+    uint256 amount;
+
+    // cache state
+    uint256 toBalanceBefore    = balanceOf(to);
+    uint256 toVotingPowerBefore = delegatedVotingPower(delegatee(to));
+    uint256 otherBalanceBefore = balanceOf(other);
+    uint256 totalSupplyBefore  = totalSupply();
+
+    // Safe require thats avoid absurd counter-examples.
+    require toVotingPowerBefore <= sumOfVotingPower;
+
+    // run transaction
+    mint@withrevert(e, to, amount);
+
+    // check outcome
+    if (lastReverted) {
+        assert e.msg.sender != owner() || to == 0 || totalSupplyBefore + amount > max_uint256 ||
+            toVotingPowerBefore + amount > max_uint256;
+    } else {
+        // updates balance and totalSupply
+        assert e.msg.sender == owner();
+        assert to_mathint(balanceOf(to)) == toBalanceBefore   + amount;
+        assert to_mathint(totalSupply()) == totalSupplyBefore + amount;
+
+        // no other balance is modified
+        assert balanceOf(other) != otherBalanceBefore => other == to;
+    }
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rules: burn behavior and side effects                                                                               │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule burn(env e) {
+    requireInvariant balancesLTEqTotalSupply();
+    assert isTotalSupplyGTEqSumOfVotingPower();
+    requireInvariant zeroAddressNoVotingPower();
+    require nonpayable(e);
+
+    address from;
+    address other;
+    uint256 amount;
+    require from == e.msg.sender;
+    // cache state
```

Comment:
Not sure if those comments are useful. Same for `run transaction` and `check outcome`

Reply:
For the record, not doing this change as we agreed to preserve simpler diffs w.r.t to OZ's original spec.